src/callbacks/ltfb.cpp

```diff
@@ -34,283 +33,551 @@
-  const auto local_trainer = comm.get_trainer_rank();
-  const std::string send_dir = (basedir
-                                + m.get_name()
-                                + "_trainer" + std::to_string(local_trainer)
-                                + "_step" + std::to_string(step));
-  const std::string recv_dir = (basedir
-                                + m.get_name()
-                                + "_trainer" + std::to_string(partner_trainer)
-                                + "_step" + std::to_string(step));
-
-  // Save model checkpoint
-  {
-    persist p;
-    p.set_cb_type(callback_type::model_only);
-    p.open_checkpoint(send_dir, comm.am_trainer_master());
+    // Checkpoint directories
+    const auto basedir = (ckpt_basedir_.empty()
+                          ? std::string("")
+                          : add_delimiter(ckpt_basedir_));
+    const auto local_trainer = comm.get_trainer_rank();
+    const std::string send_dir = (basedir
+                                  + m.get_name()
+                                  + "_trainer" + std::to_string(local_trainer)
+                                  + "_step" + std::to_string(step));
+    const std::string recv_dir = (basedir
+                                  + m.get_name()
+                                  + "_trainer" + std::to_string(partner_trainer)
+                                  + "_step" + std::to_string(step));
+
+    // Save model checkpoint
+    {
+      persist p;
+      p.set_cb_type(callback_type::model_only);
+      p.open_checkpoint(send_dir, comm.am_trainer_master());
+      comm.trainer_barrier();
+      m.save_to_checkpoint_shared(p);
+      p.close_checkpoint();
+    }
+
+    // Synchronize with partner trainer
+    comm.trainer_barrier();
+    if (comm.am_trainer_master()) {
+      int send{0}, recv{0};
+      comm.sendrecv(&send, 1, partner_trainer, 0,
+                    &recv, 1, partner_trainer, 0,
+                    El::SyncInfo<El::Device::CPU>{});
+    }
     comm.trainer_barrier();
-    m.save_to_checkpoint_shared(p);
-    p.close_checkpoint();
-  }
 
-  // Synchronize with partner trainer
-  comm.trainer_barrier();
-  if (comm.am_trainer_master()) {
-    int send{0}, recv{0};
-    comm.sendrecv(&send, 1, partner_trainer, 0,
-                  &recv, 1, partner_trainer, 0,
-                  El::SyncInfo<El::Device::CPU>{});
-  }
-  comm.trainer_barrier();
+    // Load model checkpoint from partner trainer
+    {
+      persist p;
+      p.set_cb_type(callback_type::model_only);
+      p.open_restart(recv_dir);
+      m.load_from_checkpoint_shared(p);
+      p.close_restart();
+    }
 
-  // Load model checkpoint from partner trainer
-  {
-    persist p;
-    p.set_cb_type(callback_type::model_only);
-    p.open_restart(recv_dir);
-    m.load_from_checkpoint_shared(p);
-    p.close_restart();
+    restore_model_weights(m, restore_weights);
   }
+private:
+  std::string ckpt_basedir_;
+};// class CheckpointFile
+
+class CheckpointBlob final
```

Comment:
As compelling as "blob" is, it's better to have consistent naming:
```suggestion
class CheckpointBinary final
```